sync.js

```diff
@@ -460,15 +459,23 @@ function WifiPeer (connection, info) {
   return info
 }
 
-// key is String or Buffer
-function hash (key) {
-  if (typeof key === 'string') {
-    key = Buffer.from(key, 'hex')
+/**
+ * Generate a discovery key for a given projectKey. If projectKey is undefined
+ * then it will use SYNC_DEFAULT_KEY as the discovery key (this is for backwards
+ * compatibility with clients that did not use projectKeys)
+ *
+ * @param {String|Buffer} projectKey A unique random key identifying the
+ * project. Must be 32-byte Buffer or a string hex encoding of a 32-Byte buffer
+ */
+function discoveryKey (projectKey) {
+  if (typeof projectKey === 'undefined') return SYNC_DEFAULT_KEY
```

Comment:
Having a function that "generates a discovery key for a given projectKey" fall back to returning a completely unrelated value (`SYNC_DEFAULT_KEY`) feels weird. I think keeping it outside makes this function much simpler && clearer.

Reply:
yeah, I get that. What do you think the best way of throwing is though? Test the key is valid in the join/leave functions and throw from there before trying to create a discovery key?

Reply:
The alternative I can think of is to test the project key is valid in both the join and leave functions and throw there, but then this function is just crypto.discoveryKey and we've ended up copying the same code into two places.

Reply:
I'm ok with merging as-is and coming back to this later, which we can refactor /wo breaking any API stuff. :+1: